### PR TITLE
Switch to ES6 features (close #10, #104)

### LINKED ISSUE #10: ES6 features
\cc @VasilyStrelyaev @LavrovArtem @miherlosev 

Based on https://github.com/superroma/testcafe-hammerhead/issues/5

- [ ] Use arrow functions
- [ ] Use classes instead of modules
- [ ] Rename constants in accordance with the new rules

### PR DESCRIPTION
**Pull request by churkin**
\cc @inikulin @AlexanderMoskovkin  @miherlosev

### CODE REVIEW

**Comment by inikulin** [comment]
@churkin Tests fail. BTW we've completely forgot about replacing callback-based async funcs with promises. And I can't find related issue. Do you plan to do that in context of this PR? If not, please fill separate issue, please.

**Comment by churkin** [comment]
@inikulin I have created issue "Replace callback-based async funcs with promises" #108

**Comment by churkin** [comment]
FPR

File: src/client/hammerhead.js [hunk 1]
```diff
@@ -6,111 +10,112 @@ import * as Types from './utils/types';
...
+            Types:    Types
+        };
     }
-};
 
-export function on (evtName, handler) {
-    var eventOwner = getEventOwner(evtName);
+    _getEventOwner (evtName) {
+        switch (evtName) {
+            case this.sandbox.event.unload.BEFORE_UNLOAD_EVENT:
```
Comment (inikulin) [inline src/client/hammerhead.js]: Why not use `this.EVENTS` as shortcut here?

File: src/client/hammerhead.js [hunk 2]
```diff
@@ -6,111 +10,112 @@ import * as Types from './utils/types';
...
+        if (settings)
+            Settings.set(settings);
+
+        this.sandbox.attach(this.window);
+    }
 }
 
-export var get     = require;
-export var sandbox = sandbox;
+var hammerhead = new Hammerhead(window);
```
Comment (inikulin) [inline src/client/hammerhead.js]: Shouldn't we expose hammerhead the way described in #18 ?

File: src/client/hammerhead.js [hunk 3]
```diff
@@ -1,4 +1,8 @@
 /*eslint-disable no-native-reassign*/
```
Comment (inikulin) [inline src/client/hammerhead.js]: It would be nice to implement #104 in context of this PR as well.

File: src/client/sandbox/cookie.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ import UrlUtil from '../utils/url';
 import Settings from '../settings';
 import * as Cookie from '../utils/cookie';
 import { isCrossDomainWindows } from '../utils/dom';
-import { SET_COOKIE as SET_COOKIE_CMD } from '../../service-msg-cmd';
+import { setCookie as SET_COOKIE_CMD } from '../../service-msg-cmd';
```
Comment (inikulin) [inline src/client/sandbox/cookie.js]: Better do it this way:
```js
import * as COMMAND from '../../service-msg-cmd';
```
and use it as:
```js
COMMAND.setCookie
```
It would be nice to rename `service-msg-cmd` to `commands` and move it to `session` for better semantics (this will partially solve #68)

File: src/client/sandbox/event/element-editing-watcher.js [hunk 5]
```diff
@@ -1,65 +1,65 @@
-import * as DOM from '../../utils/dom';
-import * as EventSimulator from './simulator';
 import NativeMethods from '../native-methods';
 import Const from '../../../const';
+import * as DOM from '../../utils/dom';
```
Comment (inikulin) [inline src/client/sandbox/event/element-editing-watcher.js]: Better name it `DomUtils`

File: src/client/sandbox/event/element-editing-watcher.js [hunk 6]
```diff
@@ -1,65 +1,65 @@
-import * as DOM from '../../utils/dom';
-import * as EventSimulator from './simulator';
 import NativeMethods from '../native-methods';
```
Comment (inikulin) [inline src/client/sandbox/event/element-editing-watcher.js]: We still have a huge inconsistency in naming in many places: native methods here is an instance, so it should be named `nativeMethods`. Do you mind fixing all such places in context of this PR?

File: src/client/sandbox/event/simulator.js [hunk 7]
```diff
@@ -1,520 +1,483 @@
...
-    }
+        }
 
-    //NOTE: After the MouseEvent was created by using initMouseEvent method pageX and pageY properties equal zero (only in IE9).
-    //We can set them only by defineProperty method (B253930)
-    if (Browser.isIE9) {
-        var currentDocument = DOM.findDocument(el);
-        var documentScroll  = Style.getElementScroll(currentDocument);
+        if (pointerRegExp.test(args.type) && (window.PointerEvent || window.MSPointerEvent)) {
+            var pointEvent       = Browser.version >
```
Comment (inikulin) [inline src/client/sandbox/event/simulator.js]: This browser version checks bothers me. I don't see any browser type checks above, so seems like we check **any** browser for the version. Or we have IE check in the upper condition, but it's so unobvious that I was not able to parse it. Anyway, we need to refactor this code. \cc @helen-dikareva I guess

File: src/client/sandbox/event/timeout.js [hunk 8]
```diff
@@ -1,86 +1,90 @@
+import SandboxBase from '../base';
```
Comment (inikulin) [inline src/client/sandbox/event/timeout.js]: Since it's replaces both `setTimeout` and `setInterval` seems like it's better be named `TimersSandbox`

File: src/client/sandbox/iframe.js [hunk 9]
```diff
@@ -3,41 +3,40 @@ import NativeMethods from './native-methods';
 import Settings from '../settings';
 import * as DOM from '../utils/dom';
 import { syncServiceMsg } from '../transport';
-import { isMozilla, isIE, isWebKit } from '../utils/browser';
+import { isMozilla, isWebKit } from '../utils/browser';
 import { DOM_SANDBOX_OVERRIDE_DOM_METHOD_NAME } from '../../const';
-import { GET_IFRAME_TASK_SCRIPT as GET_IFRAME_TASK_SCRIPT_CMD } from '../../service-msg-cmd';
+import { getIframeTaskScript as GET_IFRAME_TASK_SCRIPT_CMD } from '../../service-msg-cmd';
```
Comment (inikulin) [inline src/client/sandbox/iframe.js]: Same as above, with `COMMAND`

File: src/client/sandbox/iframe.js [hunk 10]
```diff
@@ -3,41 +3,40 @@ import NativeMethods from './native-methods';
...
 import { isSupportedProtocol, isIframeWithoutSrc } from '../utils/url';
 
+const IFRAME_WINDOW_INITED = 'hh_iwi_5d9138e9';
```
Comment (inikulin) [inline src/client/sandbox/iframe.js]: I think we need some common pattern for all our DOM-integrated constants, with some naming convention with clear names, because currently it's pretty random and not clear. I think we can do it in separate PR after v1.0, can you fill the issue pls?

File: src/client/sandbox/node/window.js [hunk 11]
```diff
@@ -10,7 +11,7 @@ export default class WindowSandbox extends SandboxBase {
     constructor (sandbox) {
         super(sandbox);
 
-        this.UNCAUGHT_JS_ERROR = 'uncaughtJSError';
+        this.UNCAUGHT_JS_ERROR_EVENT = 'uncaughtJSError';
```
Comment (inikulin) [inline src/client/sandbox/node/window.js]: I'm starting to wondering if it was a good idea to keep constants on instance level. On one hand we'll be able replace values for testing purposes, on other hand we will need to make instance methods which may be static but uses this constants. I'd stick with the static constants, e.g.:
```js
WindowSandbox.UNCAUGHT_JS_ERROR_EVENT = 'uncaughtJSError';
```

File: src/client/sandbox/upload/info-manager.js [hunk 12]
```diff
@@ -1,18 +1,18 @@
...
 import ServiceCommands from '../../../service-msg-cmd';
-import { SHADOW_UI_CLASSNAME_POSTFIX } from '../../../const';
 import Transport from '../../transport';
 import Settings from '../../settings';
 import FileListWrapper from './file-list-wrapper';
+import * as Browser from '../../utils/browser';
+import * as HiddenInfo from './hidden-info';
+import { SHADOW_UI_CLASSNAME_POSTFIX } from '../../../const';
+
+const FAKE_PATH_STRING         = 'C:\\fakepath\\';
```
Comment (inikulin) [inline src/client/sandbox/upload/info-manager.js]: Just wondering why we have this Windows-specific code in production and does it even work on other OS?

File: src/client/transport.js [hunk 13]
```diff
@@ -1,245 +1,240 @@
...
-}, true);
+            //NOTE: if unloading was canceled switch back to async XHR
+            NativeMethods.setTimeout.call(window, () => this.useAsyncXhr = true, this.SWITCH_BACK_TO_ASYNC_XHR_DELAY);
+        }, true);
+    }
 
-Transport.sendNextQueuedMsg = function (queueId) {
-    var queueItem = msgQueue[queueId][0];
+    static _createXMLHttpRequest (async) {
+        var xhr = new NativeMethods.XMLHttpRequest();
```
Comment (inikulin) [inline src/client/transport.js]: We have a `SERVICE_MSG_REQUEST_FLAG` in https://github.com/churkin/testcafe-hammerhead/blob/es6/src/client/sandbox/xhr/index.js . It was used to skip tracking of the transport messages. I'm wondering, since we use native methods now, do we still need this flag?

**Comment by inikulin** [comment]
\r-

File: src/client/hammerhead.js (hunk 2 above)
Comment (churkin) [inline src/client/hammerhead.js]: Yes, but not at this stage

File: src/client/hammerhead.js (hunk 3 above)
Comment (churkin) [inline src/client/hammerhead.js]: As a separate commit?

File: src/client/hammerhead.js (hunk 2 above)
Comment (inikulin) [inline src/client/hammerhead.js]: It would be nice to do that ASAP. It's blocking for the NPM publishing.

File: src/client/hammerhead.js (hunk 3 above)
Comment (inikulin) [inline src/client/hammerhead.js]: As you wish

File: src/client/sandbox/iframe.js (hunk 10 above)
Comment (churkin) [inline src/client/sandbox/iframe.js]: Get all constant values to a identical type #111

File: src/client/sandbox/node/window.js (hunk 11 above)
Comment (churkin) [inline src/client/sandbox/node/window.js]: Move class constants to the prototype #112

File: src/client/transport.js (hunk 13 above)
Comment (churkin) [inline src/client/transport.js]: Use the native XMLHttpRequest instead of a SERVICE_MSG_REQUEST_FLAG flag mechanism #113

File: src/client/hammerhead.js (hunk 1 above)
Comment (churkin) [inline src/client/hammerhead.js]: done

File: src/client/sandbox/cookie.js (hunk 4 above)
Comment (churkin) [inline src/client/sandbox/cookie.js]: done

File: src/client/sandbox/event/simulator.js (hunk 7 above)
Comment (churkin) [inline src/client/sandbox/event/simulator.js]: done

File: src/client/sandbox/event/timeout.js (hunk 8 above)
Comment (churkin) [inline src/client/sandbox/event/timeout.js]: done

File: src/client/sandbox/iframe.js (hunk 9 above)
Comment (churkin) [inline src/client/sandbox/iframe.js]: done

File: src/client/sandbox/upload/info-manager.js (hunk 12 above)
Comment (churkin) [inline src/client/sandbox/upload/info-manager.js]: How uploading works on the Linux/MacOS platforms? #115

File: src/client/sandbox/event/element-editing-watcher.js (hunk 5 above)
Comment (churkin) [inline src/client/sandbox/event/element-editing-watcher.js]: done

File: src/client/sandbox/event/element-editing-watcher.js (hunk 6 above)
Comment (churkin) [inline src/client/sandbox/event/element-editing-watcher.js]: done

File: src/client/sandbox/upload/info-manager.js (hunk 12 above)
Comment (churkin) [inline src/client/sandbox/upload/info-manager.js]: How uploading works on the Linux/MacOS platforms? #115

**Comment by churkin** [comment]
FPR

File: src/client/sandbox/event/index.js [hunk 14]
```diff
@@ -1,210 +1,235 @@
...
+        this._createInternalHandlers();
+    }
+
+    _createOverridedMethods () {
+        var selection        = this.selection;
+        var focusBlurSandbox = this.focusBlur;
+        var listeners        = this.listeners;
+        var eventSimulator   = this.eventSimulator;
+
+        this.overridedDispatchEvent = function (ev) {
```
Comment (inikulin) [inline src/client/sandbox/event/index.js]: It would be nice to have these methods compound into the object, e.g.: `overridedMethods.dispatchEvent`.
It will give us 2 benefits:
1)We will get rid of `overrided` prefix.
2)`this` inside them will not be so confusing. Currently it's intended to the element, but from the code it's looks like `this` points to `EventSandbox`

**Comment by inikulin** [comment]
Great job. Just a single minor issue. Everything else looks pretty good for (but I would like more people to keep an eye on it, since it's huge \cc @helen-dikareva @miherlosev @AndreyBelym ?).

**Comment by inikulin** [comment]
\r-

**Comment by helen-dikareva** [comment]
Maybe it makes sense to send 'window' like parameter in Hammerhead init()

**Comment by inikulin** [comment]
BTW, why we need `init` method in Hammerhead?

File: src/client/sandbox/event/index.js (hunk 14 above)
Comment (churkin) [inline src/client/sandbox/event/index.js]: done

**Comment by churkin** [comment]
done
@helen-dikareva 
>Maybe it makes sense to send 'window' like parameter in Hammerhead init()

**Comment by helen-dikareva** [comment]
src/client/sandbox/event/focus-blur.js
https://github.com/superroma/testcafe-hammerhead/pull/106/files#diff-5a4f875dc63f1cf4193b951060ba8bfdR4
Use instead of
<code>
import * as Style from '../../utils/style';
</code>
->
<code>
import * as styleUtils from '../../utils/style';
</code>

and etc. Check everywhere.

**Comment by churkin** [comment]
@helen-dikareva  done

**Comment by miherlosev** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by AndreyBelym** [comment]
lgtm

**Comment by inikulin** [comment]
lgtm
